text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+
+Extensions are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
+However, instead of a `buildpack.toml` file, extensions have a `extension.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[extension]
+id = "<extension ID>"
+name = "<extension name>"
+version = "<extension version>"
+homepage = "<extension homepage>"
+description = "<extension description>"
+keywords = [ "<string>" ]
+
+[[extension.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Extensions may be packaged and examined similar to buildpacks, but with analogous `pack extension` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for extensions.
+
+Unlike buildpacks,
+- Extensions must not be included in a meta-buildpacks
+- Extensions must not have `order`/`group` definitions in `extension.toml`
+
+Extensions participate in the buildpack detection process, with the same UID, GID, and interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for extensions, and they are assumed to pass detection when it is not present. Just like with buildpacks, a /bin/detect that exits with a 0 exit code passes detection, and fails otherwise.
+- Extensions may only output `provides` entries to the build plan. They must not output `requires`.
+- Extensions must all precede regular buildpacks in `order` definitions (e.g., in `builder.toml`).
+- Extensions are always `optional`.
+
+Extensions generate Dockerfiles before the regular buildpack build phase.
+To generate these Dockerfiles, the lifecycle executes the extension's `/bin/build` executable with the same UID, GID, and interface as regular buildpacks.
+However,
+- Extensions `/bin/build` must not write to the app directory.
+- Extensions `/bin/build` may be executed in parallel.
+- Extensions `<layers>` directory is replaced by an `<output>` directory.
+- If an extension is missing `/bin/build`, the extension root is treated as a pre-populated `<output>` directory.
+
+After `/bin/build` executes, the `<output>` directory may contain
+- `build.toml`, with the same contents as a normal buildpack's `build.toml`, but
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `build.Dockerfile` or `Dockerfile`
+- `launch.toml`, with the same contents as a normal buildpack's `launch.toml`, but
+  - Without the `processes` table array
+  - Without the `slices` table array
+  - With an additional `args` table array with `name` and `value` fields that are provided as build args to `run.Dockerfile` or `Dockerfile`
+
+- Either `Dockerfile` or either or both of `build.Dockerfile` and `run.Dockerfile`
+
+Support for other instruction formats, e.g., LLB JSON files, could be added in the future.
+
+`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively.
```

Comment:
```suggestion
`build.Dockerfile`, `run.Dockerfile`, and `Dockerfile` target the builder image, runtime base image, or both base images, respectively. Dockerfiles that target the build image must begin with the following lines: `ARG base_image\nFROM ${base_image}`
```

Reply:
Hmm - how would this work with multi-stage Dockerfiles? 

```
FROM some-registry/repo/image as external_dep

FROM ${base_image}
COPY --from=external_dep ...
```

Reply:
I think we need to disallow multi-stage and essentially ignore `FROM` clauses and that we execute the commands directly on the running build container.

Reply:
Hmm I don't follow - the example you gave doesn't start with 
```
ARG base_image
FROM ${base_image}
```
so I think it would be disallowed.

Reply:
I'm thinking about multi-stage builds and maybe we should explicitly disallow them for `build` exts or else we need to parse the `Dockerfile` and understand the last stage to be `base_image`, right?

This would pass the validation described, but still ultimately ends up trying to use a different build base image.

```
ARG base_image
FROM ${base_image}
# do nothing to pass validation

FROM some-registry/repo/image as external_dep

FROM ${base_image}
COPY --from=external_dep ...
RUN some-command

FROM alpine
RUN echo hmm
```